### PR TITLE
TC : Search and Add product

### PR DESCRIPTION
**Pull request by NFartenadze**

### CODE REVIEW

**Review by FCosta20 — CHANGES_REQUESTED** [review]
All your tests are failing, please check:

Screenshot 2023-12-28 at 11 56 33

**Comment by NFartenadze** [comment]
i have 2 common steps in this folder, i dont know if that is a good way to handle it

File: cypress/e2e/features/unauthorised.feature [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    When the user searches for HTC One M
+    Then the search results should contain products with the title HTC One M
+    When the user clicks on HTC One M
+    Then the user should be navigated to product detail page
+    Then the product details should match expected values
+   
+
+
+
+  Scenario: Check if user can view details of product V2
```
Comment (FCosta20) [inline cypress/e2e/features/unauthorised.feature]: Why do you need 2 versions of the same scenario?

File: cypress/e2e/features/unauthorised.feature [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  Background: 
+    Given the user visits the nopcommerce website
+
+  Scenario: Search product and add to cart
+    When the user searches for Nikon
+    Then the search results should contain product with the title Nikon
+    When the user clicks on searched product
+    Then the user should be navigated to product detail page
+    When the user adds product to cart
+    Then the cart quantity should be updated to 1
```
Comment (FCosta20) [inline cypress/e2e/features/unauthorised.feature]: Maybe add a step to go to cart page and verify that the product is in cart page?

File: cypress/e2e/features/authorised.feature
Comment (FCosta20) [inline cypress/e2e/features/authorised.feature:1]: Why is this file empty?

File: cypress/support/step_definitions/authorised/CheckoutProductSteps.ts [hunk 3]
```diff
@@ -0,0 +1,74 @@
+import { When } from "@badeball/cypress-cucumber-preprocessor";
+
+const values = {
```
Comment (FCosta20) [inline cypress/support/step_definitions/authorised/CheckoutProductSteps.ts]: Use Cypress Fixtures for test data

File: cypress/support/step_definitions/authorised/CheckoutProductSteps.ts
Comment (FCosta20) [inline cypress/support/step_definitions/authorised/CheckoutProductSteps.ts:1]: Test implementations should be in e2e folder

File: cypress/support/step_definitions/authorised/CheckoutProductSteps.ts [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+  cy.get("#product-details-form").then((form) => {
+    expect(form).to.exist;
+    cy.wrap(form).find(".add-to-cart-panel button").first().click();
+    cy.wait(3000);
+    cy.get(".cart-qty").should("contain", "1");
+  });
+}
+
+When("", () => {
```
Comment (FCosta20) [inline cypress/support/step_definitions/authorised/CheckoutProductSteps.ts]: Why is this step not having any name?

File: cypress/support/step_definitions/authorised/CheckoutProductSteps.ts [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    cy.wrap(input).type(values.register[index]);
+  });
+
+  //select date
+  cy.get("select[name='DateOfBirthDay']").select("8");
+  cy.get("select[name='DateOfBirthMonth']").select("September");
+  cy.get("select[name='DateOfBirthYear']").select("2002");
+
+  cy.get("#register-button").click();
+  cy.wait(2000);
```
Comment (FCosta20) [inline cypress/support/step_definitions/authorised/CheckoutProductSteps.ts]: Please try not to hardcode waits, you can implement cypress-wait-until plugin if necessary

File: cypress/support/step_definitions/unauthorised/SearchAndAddToCartSteps.ts [hunk 6]
```diff
@@ -0,0 +1,21 @@
+import { When, Then } from "@badeball/cypress-cucumber-preprocessor";
+
+When("the user searches for Nikon", () => {
```
Comment (FCosta20) [inline cypress/support/step_definitions/unauthorised/SearchAndAddToCartSteps.ts]: Try using variables, you have the same step but for a different product

File: cypress/support/step_definitions/unauthorised/ViewProductDetailsSteps.ts [hunk 7]
```diff
@@ -0,0 +1,43 @@
+import { When, Then, DataTable } from "@badeball/cypress-cucumber-preprocessor";
+
+When("the user searches for HTC One M", () => {
+  cy.get("#small-searchterms").type("HTC One M{enter}");
+});
+
+Then(
+  "the search results should contain products with the title HTC One M",
```
Comment (FCosta20) [inline cypress/support/step_definitions/unauthorised/ViewProductDetailsSteps.ts]: Don't use this unnecessary spaces, this makes the code less readable

File: cypress/support/step_definitions/unauthorised/ViewProductDetailsV2Steps.ts [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    locator: ".additional-details",
+    expectedText: "SKU: M8_HTC_5L",
+  },
+  productPrice: {
+    locator: ".product-price",
+    expectedText: " $245.00 ",
+  },
+};
+
+import { When, Then } from "@badeball/cypress-cucumber-preprocessor";
```
Comment (FCosta20) [inline cypress/support/step_definitions/unauthorised/ViewProductDetailsV2Steps.ts]: Imports should be on top of the file, this makes the code more readable

File: cypress/e2e/features/unauthorised.feature
Comment (FCosta20) [inline cypress/e2e/features/unauthorised.feature:1]: The feature file name is not making sense with the scenarios inside the file. Please modify it so the scenarios matches the naming of the file.

**Review by FCosta20 — CHANGES_REQUESTED** [review]
Please check comments and fix in all files.

File: cypress/e2e/features/unauthorised.feature (hunk 1 above)
Comment (NFartenadze) [reply cypress/e2e/features/unauthorised.feature]: If you could check steps for both of them, I used different ways to check if details are present. but i ll remove second one in new commit.

File: cypress/e2e/features/authorised.feature
Comment (NFartenadze) [reply cypress/e2e/features/authorised.feature:1]: i added that file accidentally, i am planning to write test which requires authorisation, thus i divided feature files in two

File: cypress/support/step_definitions/authorised/CheckoutProductSteps.ts
Comment (NFartenadze) [reply cypress/support/step_definitions/authorised/CheckoutProductSteps.ts:1]: okay i will move them, i just searched for conventions and how people structure them, followed those

**Comment by NFartenadze** [comment]
i updated based on your comments, i just have one problem, if i put step definitions into folder it cannot find them, ill work on that

**Comment by NFartenadze** [comment]
after authorizing website logs me out when i click checkout, so i wrote checkout as guest

File: cypress/support/commands.ts [hunk 9]
```diff
@@ -1,37 +1,39 @@
...
+// import * as registerValues from "../fixtures/registrationFields.json";
+
+declare namespace Cypress {
+  interface Chainable<Subject = any> {
+    login(): Chainable<Subject>;
+    register(): Chainable<Subject>;
+  }
+}
+
+// Cypress.Commands.add("register", () => {
```
Comment (FCosta20) [inline cypress/support/commands.ts]: Please don't push commented or not necessary code, uncomment if used or delete if not used.

File: cypress/support/e2e.ts [hunk 10]
```diff
@@ -23,3 +23,6 @@ Cypress.on("uncaught:exception", (err, runnable) => {
   // returning false here prevents Cypress from failing the test
   return false;
 });
+
```
Comment (FCosta20) [inline cypress/support/e2e.ts:26]: Why are you adding just empty spaces?

File: cypress/fixtures/registrationFields.json
Comment (FCosta20) [inline cypress/fixtures/registrationFields.json:1]: This file is not used, why do you need it?

File: cypress/fixtures/navigationCategories.json [hunk 11]
```diff
@@ -0,0 +1,9 @@
+[
```
Comment (FCosta20) [inline cypress/fixtures/navigationCategories.json]: This is not a json, this is just an array, no need to have a whole file for this, you can add this array into another json or if this json will be bigger you can add this array into a json inside this file

File: cypress/support/commands.ts [hunk 12]
```diff
@@ -1,37 +1,39 @@
...
-// declare global {
-//   namespace Cypress {
-//     interface Chainable {
-//       login(email: string, password: string): Chainable<void>
-//       drag(subject: string, options?: Partial<TypeOptions>): Chainable<Element>
-//       dismiss(subject: string, options?: Partial<TypeOptions>): Chainable<Element>
-//       visit(originalFn: CommandOriginalFn, url: string, options: Partial<VisitOptions>): Chainable<Element>
+// import * as registerValues from "../fixtures/registrationFields.json";
+
+declare namespace Cypress {
```
Comment (FCosta20) [inline cypress/support/commands.ts]: Can you explain why you need this?

**Review by FCosta20 — CHANGES_REQUESTED** [review]
You are writing similar scenarios in separated feature files. You can create 1 feature file called ProductTests, ProductScenarios, Products, etc. and just write the 3 scenarios there, also you can reuse some of the steps this way

File: cypress/fixtures/navigationCategories.json (hunk 11 above)
Comment (NFartenadze) [reply cypress/fixtures/navigationCategories.json]: okay, i just separated them based on what they were used for

File: cypress/support/commands.ts (hunk 12 above)
Comment (NFartenadze) [reply cypress/support/commands.ts]: i had steps which were similar in file commonSteps.ts but cypress gave me error that it couldnt find step definitions, so i had to add them back to each tests they were used in, like ,,user navigates to nopcommerce''

**Review by FCosta20 — APPROVED** [review]

**Pull request merged**